File: C++/queue.cpp
#include "queue.h"

#include <istream>
#include <ostream>
#include <utility>
#include <vector>

namespace {
void write_string_bin(std::ostream& os, const std::string& s) {
    const std::uint32_t n = static_cast<std::uint32_t>(s.size());
    os.write(reinterpret_cast<const char*>(&n), sizeof(n));
    os.write(s.data(), static_cast<std::streamsize>(n));
}
std::string read_string_bin(std::istream& is) {
    std::uint32_t n = 0;
    is.read(reinterpret_cast<char*>(&n), sizeof(n));
    std::string s(n, '\0');
    if (n > 0) {
        is.read(&s[0], static_cast<std::streamsize>(n));
    }
    return s;
}
}

Queue::Queue(std::string name) : name(std::move(name)) {}

Queue::~Queue() { clear(); }

void Queue::clear() {
    Node* cur = head_;
    while (cur) {
        Node* nxt = cur->next;
        delete cur;
        cur = nxt;
    }
    head_ = nullptr;
    tail_ = nullptr;
}

void Queue::push(const std::string& data) {
    auto* n = new Node(data);
    if (!tail_) {
        head_ = tail_ = n;
        return;
    }
    tail_->next = n;
    tail_ = n;
}

std::string Queue::pop() {
    if (!head_) {
        return "";
    }
    Node* del = head_;
    std::string out = del->data;
    head_ = head_->next;
    if (!head_) {
        tail_ = nullptr;
    }
    delete del;
    return out;
}

std::string Queue::peek() const { return head_ ? head_->data : ""; }

bool Queue::is_empty() const { return head_ == nullptr; }

std::string Queue::read() const {
    std::string out;
    bool first = true;
    for (Node* cur = head_; cur; cur = cur->next) {
        if (!first) {
            out += ' ';
        }
        first = false;
        out += cur->data;
    }
    return out;
}

void Queue::serialize_binary(std::ostream& os) const {
    std::vector<std::string> values;
    for (Node* cur = head_; cur; cur = cur->next) {
        values.push_back(cur->data);
    }
    const std::uint32_t sz = static_cast<std::uint32_t>(values.size());
    os.write(reinterpret_cast<const char*>(&sz), sizeof(sz));
    for (const auto& v : values) {
        write_string_bin(os, v);
    }
}

void Queue::deserialize_binary(std::istream& is) {
    clear();
    std::uint32_t sz = 0;
    is.read(reinterpret_cast<char*>(&sz), sizeof(sz));
    for (std::uint32_t i = 0; i < sz; ++i) {
        push(read_string_bin(is));
    }
}

void Queue::serialize_text(std::ostream& os) const {
    std::vector<std::string> values;
    for (Node* cur = head_; cur; cur = cur->next) {
        values.push_back(cur->data);
    }
    os << values.size() << "\n";
    for (const auto& v : values) {
        os << v << "\n";
    }
}

void Queue::deserialize_text(std::istream& is) {
    clear();
    int sz = 0;
    if (!(is >> sz)) {
        return;
    }
    std::string dummy;
    std::getline(is, dummy);
    for (int i = 0; i < sz; ++i) {
        std::string line;
        std::getline(is, line);
        push(line);
    }
}

File: datastructures-go/queue.go
package datastructures

import (
	"encoding/binary"
	"fmt"
	"io"
	"strings"
)

type queueNode struct {
	data string
	next *queueNode
}

type Queue struct {
	Name string
	head *queueNode
	tail *queueNode
}

func NewQueue(name string) *Queue {
	return &Queue{Name: name}
}

func (q *Queue) clear() {
	q.head = nil
	q.tail = nil
}

func (q *Queue) Push(data string) {
	node := &queueNode{data: data}
	if q.tail == nil {
		q.head = node
		q.tail = node
		return
	}
	q.tail.next = node
	q.tail = node
}

func (q *Queue) Pop() string {
	if q.head == nil {
		return ""
	}
	data := q.head.data
	q.head = q.head.next
	if q.head == nil {
		q.tail = nil
	}
	return data
}

func (q *Queue) Peek() string {
	if q.head == nil {
		return ""
	}
	return q.head.data
}

func (q *Queue) IsEmpty() bool {
	return q.head == nil
}

func (q *Queue) Read() string {
	var values []string
	current := q.head
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	return strings.Join(values, " ")
}

func (q *Queue) SerializeBinary(w io.Writer) error {
	var values []string
	current := q.head
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	
	length := uint32(len(values))
	if err := binary.Write(w, binary.LittleEndian, length); err != nil {
		return err
	}
	
	for _, v := range values {
		strLen := uint32(len(v))
		if err := binary.Write(w, binary.LittleEndian, strLen); err != nil {
			return err
		}
		if _, err := w.Write([]byte(v)); err != nil {
			return err
		}
	}
	return nil
}

func (q *Queue) DeserializeBinary(r io.Reader) error {
	q.clear()
	
	var length uint32
	if err := binary.Read(r, binary.LittleEndian, &length); err != nil {
		return err
	}
	
	for i := uint32(0); i < length; i++ {
		var strLen uint32
		if err := binary.Read(r, binary.LittleEndian, &strLen); err != nil {
			return err
		}
		
		buf := make([]byte, strLen)
		if _, err := io.ReadFull(r, buf); err != nil {
			return err
		}
		q.Push(string(buf))
	}
	return nil
}

func (q *Queue) SerializeText(w io.Writer) error {
	var values []string
	current := q.head
	for current != nil {
		values = append(values, current.data)
		current = current.next
	}
	
	if _, err := fmt.Fprintf(w, "%d\n", len(values)); err != nil {
		return err
	}
	
	for _, v := range values {
		if _, err := fmt.Fprintf(w, "%s\n", v); err != nil {
			return err
		}
	}
	return nil
}

func (q *Queue) DeserializeText(r io.Reader) error {
	q.clear()
	
	var length int
	if _, err := fmt.Fscanf(r, "%d\n", &length); err != nil {
		return err
	}
	
	for i := 0; i < length; i++ {
		var s string
		if _, err := fmt.Fscanf(r, "%s\n", &s); err != nil {
			return err
		}
		q.Push(s)
	}
	return nil
}
